src/builder/QuarkOperationHelper.sol

```diff
@@ -0,0 +1,150 @@
+// SPDX-License-Identifier: BSD-3-Clause
+pragma solidity ^0.8.23;
+
+import {IQuarkWallet} from "quark-core/src/interfaces/IQuarkWallet.sol";
+
+import {Multicall} from "../Multicall.sol";
+
+import {Actions} from "./Actions.sol";
+import {CodeJarHelper} from "./CodeJarHelper.sol";
+import {PaycallWrapper} from "./PaycallWrapper.sol";
+import {PaymentInfo} from "./PaymentInfo.sol";
+import {QuotecallWrapper} from "./QuotecallWrapper.sol";
+
+// Helper library to merge QuarkOperations on the same chain
+library QuarkOperationHelper {
+    function mergeSameChainOperations(
+        IQuarkWallet.QuarkOperation[] memory quarkOperations,
+        Actions.Action[] memory actions
+    ) internal pure returns (IQuarkWallet.QuarkOperation[] memory, Actions.Action[] memory) {
+        // Note: Assumes quarkOperations and actions have the same length
+
+        // Arrays to keep track of unique chain IDs and their operations
+        uint256[] memory uniqueChainIds = new uint256[](quarkOperations.length);
```

Comment:
Part of me feel like if we want to have another helper lib something like `HashMap.sol` implementation for this. (something like List.sol)

Reply:
Yeah, we can consider that for a future PR